ilastik/applets/batchProcessing/batchProcessingGui.py

```diff
@@ -18,22 +18,117 @@
+    QApplication, QHBoxLayout, QLabel, QListWidget, QMessageBox, QPushButton,
+    QSizePolicy, QSpacerItem, QTabWidget, QVBoxLayout, QWidget,
+)
 
 from lazyflow.request import Request
 from volumina.utility import PreferencesManager
 from ilastik.utility import log_exception
 from ilastik.utility.gui import ThreadRouter, threadRouted
-from ilastik.applets.dataSelection.dataSelectionGui import DataSelectionGui # We borrow the file selection window function.
+from ilastik.applets.dataSelection.dataSelectionGui import DataSelectionGui  # We borrow the file selection window function.
+
+
+logger = logging.getLogger(__name__)
+
+
+class BatchProcessingDataConstraintException(Exception):
+    pass
+
+
+class FileListWidget(QListWidget):
+    """QListWidget with custom drag-n-drop for file paths
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    # Drag-and-drop
+    def dropEvent(self, dropEvent):
+        urls = dropEvent.mimeData().urls()
+        filepaths = list(map(QUrl.toLocalFile, urls))
+        filepaths = list(map(str, filepaths))
+        self.clear()
+        self.addItems(qurl.path() for qurl in urls)
+
+    def dragEnterEvent(self, event):
+        # Only accept drag-and-drop events that consist of urls to local files.
+        if not event.mimeData().hasUrls():
+            return
+        urls = event.mimeData().urls()
+        if all(map(QUrl.isLocalFile, urls)):
+            event.acceptProposedAction()
+
+    def dragMoveEvent(self, event):
+        # Must override this or else the QTableView base class steals dropEvents from us.
+        pass
+
+
+class BatchRoleWidget(QWidget):
+    """Container Widget for Batch File list and buttons
+    """
+    def __init__(self, role_name: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._role_name = role_name
+        self._init_ui()
+
+    def _init_ui(self):
+        self.select_button = QPushButton(f"Select {self._role_name} Files...")
+        self.clear_button = QPushButton(f"Clear {self._role_name} Files")
+        button_layout = QHBoxLayout()
+        button_layout.addWidget(self.select_button)
+        button_layout.addSpacerItem(QSpacerItem(0, 0, hPolicy=QSizePolicy.Expanding))
+        button_layout.addWidget(self.clear_button)
+        button_layout.setContentsMargins(0, 0, 0, 0)
+
+        self.list_widget = FileListWidget(parent=self)
+        self.list_widget.setSizePolicy(QSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding))
+        self.list_widget.setAcceptDrops(True)
+
+        main_layout = QVBoxLayout()
+        main_layout.setContentsMargins(0, 0, 0, 0)
+        main_layout.addLayout(button_layout)
+        main_layout.addWidget(self.list_widget)
+
+        self.clear_button.clicked.connect(self.clear)
+        self.select_button.clicked.connect(self.select_files)
+
+        self.setLayout(main_layout)
+
+    def get_all_item_strings(self) -> typing.List[str]:
```

Comment:
Maybe use shorter and more specific name (and turn it into a property or generator)?

```suggestion
    @property
    def filepaths(self) -> typing.List[str]:
```